### PR TITLE
Fix dates

### PR DESCRIPTION
**Pull request by eakmanrq**

### CODE REVIEW

File: sqlmesh/core/plan/definition.py [hunk 1]
```diff
@@ -162,14 +167,29 @@ def set_start(self, new_start: TimeLike) -> None:
     def end(self) -> TimeLike:
         """Returns the end of the plan or now."""
         if not self._end or not self.override_end:
-            if self.missing_intervals:
-                return max(
-                    end
-                    for intervals_per_model in self._missing_intervals.values()
+            if self._missing_intervals:
+                end, snapshot_version = max(
```
Comment (tobymao) [inline sqlmesh/core/plan/definition.py:171]: delete this section

File: sqlmesh/core/plan/definition.py [hunk 2]
```diff
@@ -162,14 +167,29 @@ def set_start(self, new_start: TimeLike) -> None:
...
+                    (end, snapshot_version)
+                    for snapshot_version, intervals_per_model in self._missing_intervals.items()
                     for _, end in intervals_per_model
                 )
+                end_and_interval: t.Tuple[int, t.Optional[IntervalUnit]] = (-1, None)
```
Comment (tobymao) [inline sqlmesh/core/plan/definition.py:171]: interval_and_units = [
(end, snapshot.model.interval_unit)
for snapshot in self.snapshots
for _, end in self._missing_intervals[snapshot.version_get_or_generate()]
]

... max(interval_and_units, key=lambda x: x[0])

File: sqlmesh/core/scheduler.py [hunk 3]
```diff
@@ -364,6 +367,44 @@ def start_date(
     return earliest
 
 
+def end_date(
+    snapshot: Snapshot, snapshots: t.Dict[SnapshotId, Snapshot] | t.Iterable[Snapshot]
+) -> t.Optional[datetime]:
+    """Get the effective/inferred end date for a snapshot.
+
+    Not all snapshots define a start date. In those cases, the model's start date
```
Comment (tobymao) [inline sqlmesh/core/scheduler.py:375]: comment is wrong

File: sqlmesh/core/scheduler.py [hunk 4]
```diff
@@ -364,6 +367,44 @@ def start_date(
...
+
+    Args:
+        snapshot: snapshot to infer start date.
+        snapshots: a catalog of available snapshots.
+
+    Returns:
+        Start datetime object.
+    """
+
+    if snapshot.model.start:
```
Comment (tobymao) [inline sqlmesh/core/scheduler.py:386]: why is this dealing with start dates?

File: sqlmesh/core/scheduler.py [hunk 5]
```diff
@@ -377,6 +418,26 @@ def earliest_start_date(snapshots: t.Iterable[Snapshot]) -> datetime:
     return yesterday()
 
 
+def latest_end_date(snapshots: t.Iterable[Snapshot]) -> datetime | date:
```
Comment (tobymao) [inline sqlmesh/core/scheduler.py:421]: this can probably be dried up , looks like there's duplicate logic

File: sqlmesh/core/scheduler.py [hunk 6]
```diff
@@ -377,6 +380,26 @@ def earliest_start_date(snapshots: t.Iterable[Snapshot]) -> datetime:
...
+    """Get the latest end date from a collection of snapshots.
+
+    Args:
+        snapshots: Snapshots to find latest end date.
+    Returns:
+        The latest end date or now if none is found."""
```
Comment (vchan) [inline sqlmesh/core/scheduler.py:389]: Nit: `"""` usually goes on its own line

File: sqlmesh/core/scheduler.py [hunk 7]
```diff
@@ -377,6 +380,26 @@ def earliest_start_date(snapshots: t.Iterable[Snapshot]) -> datetime:
...
+    snapshots = list(snapshots)
+    if not any(snapshot.intervals for snapshot in snapshots):
+        return now()
+    if snapshots:
+        end_date, interval = max(
+            (snapshot.intervals[-1][1], snapshot.model.interval_unit()) for snapshot in snapshots
```
Comment (tobymao) [inline sqlmesh/core/scheduler.py:395]: if snapshot.intervals

File: sqlmesh/core/scheduler.py [hunk 8]
```diff
@@ -377,6 +380,26 @@ def earliest_start_date(snapshots: t.Iterable[Snapshot]) -> datetime:
...
 
+def latest_end_date(snapshots: t.Iterable[Snapshot]) -> datetime | date:
+    """Get the latest end date from a collection of snapshots.
+
+    Args:
+        snapshots: Snapshots to find latest end date.
+    Returns:
+        The latest end date or now if none is found."""
+    snapshots = list(snapshots)
+    if not any(snapshot.intervals for snapshot in snapshots):
```
Comment (tobymao) [inline sqlmesh/core/scheduler.py:391]: delete this section

File: sqlmesh/core/scheduler.py (hunk 7 above)
Comment (tobymao) [reply sqlmesh/core/scheduler.py:395]: store the intervals_and_units, and check that there are none here and return now() as opposed to 391

File: tests/core/test_plan.py [hunk 9]
```diff
@@ -411,4 +413,4 @@ def test_end_from_missing_instead_of_now(make_snapshot, mocker: MockerFixture):
     plan = Plan(context_diff_mock, dag, state_reader_mock, is_dev=True)
 
     assert plan.start == to_timestamp(expected_start)
-    assert plan.end == to_timestamp(expected_end)
+    assert plan.end == to_date(expected_end) - timedelta(days=1)
```
Comment (tobymao) [inline tests/core/test_plan.py:416]: why was it needed?

**Review by tobymao — APPROVED** [review]

**Pull request merged**